drivers/misc/bcm2835_smi.c

```diff
@@ -0,0 +1,1108 @@
+	inst->dma_config.dst_addr_width = DMA_SLAVE_BUSWIDTH_4_BYTES;
+	/* DREQ channel: SMI */
+	inst->dma_config.slave_id = 4 | (15 << 28) | (8 << 24);
+	inst->dma_config.src_addr = inst->smi_regs_busaddr
+		+ offsetof(struct smi_register_map, SMID);
+	inst->dma_config.dst_addr = inst->dma_config.src_addr;
+	/* Direction unimportant - always reconfigured before use */
+	inst->dma_config.direction = DMA_DEV_TO_MEM;
+	dmaengine_slave_config(inst->dma_chan, &inst->dma_config);
+	/* Alloc and map bounce buffers */
+	for (i = 0; i < DMA_BOUNCE_BUFFER_COUNT; ++i) {
+		inst->bounce.buffer[i] =
+		dma_alloc_coherent(inst->dev, DMA_BOUNCE_BUFFER_SIZE,
+				&inst->bounce.phys[i],
+				GFP_KERNEL);
+		if (!inst->bounce.buffer[i]) {
+			dev_err(inst->dev, "Could not allocate buffer!");
+			alloc_failed = true;
+			rv = -ENOMEM;
+			break;
+		}
+		inst->bounce.sgl[i] = smi_alloc_scatterlist_from_buffer
+		(
+			inst,
+			inst->bounce.phys[i],
+			DMA_BOUNCE_BUFFER_SIZE
+		);
+	}
+
+	if (alloc_failed) {
+		for (i = 0; i < DMA_BOUNCE_BUFFER_COUNT; ++i) {
+			if (!inst->bounce.buffer[i])
+				break;
+			dma_free_coherent(inst->dev, DMA_BOUNCE_BUFFER_SIZE,
+				inst->bounce.buffer[i], inst->bounce.phys[i]);
+			kfree(inst->bounce.sgl[i]);
+			inst->bounce.buffer[i] = NULL;
+			inst->bounce.sgl[i] = NULL;
+		}
+	}
+
+	return alloc_failed;
+}
+
+static int bcm2835_smi_probe(struct platform_device *pdev)
+{
+	int err;
+	struct device *dev = &pdev->dev;
+	struct device_node *node = dev->of_node;
+	struct resource *ioresource;
+	struct bcm2835_smi_instance *inst;
+
+	/* Allocate buffers and instance data */
+
+	inst = kzalloc(sizeof(struct bcm2835_smi_instance), GFP_KERNEL);
+
+	if (!inst) {
+		err = -ENOMEM;
+		goto failed_inst_alloc;
+	}
+
+	inst->buffer_sgl = kzalloc(sizeof(struct scatterlist), GFP_KERNEL);
+
+	if (!inst->buffer_sgl) {
+		err = -ENOMEM;
+		goto failed_sgl_alloc;
+	}
+
+	inst->dev = dev;
+
+	/* Allocate resources from device tree if available,
+	   or use hardcoded addresses if not. */
+
+	if (node) {
+		ioresource = platform_get_resource(pdev, IORESOURCE_MEM, 0);
+		inst->smi_regs_ptr = devm_ioremap_resource(dev, ioresource);
+		ioresource = platform_get_resource(pdev, IORESOURCE_MEM, 1);
+		inst->cm_smi_regs_ptr = devm_ioremap_resource(dev, ioresource);
+		inst->smi_regs_busaddr = be32_to_cpu(
+			*of_get_address(node, 0, NULL, NULL));
+		of_property_read_u32(node,
+				     "brcm,smi-clock-source",
+				     &inst->clock_source);
+		of_property_read_u32(node,
+				     "brcm,smi-clock-divisor",
+				     &inst->clock_divisor);
+	} else {
+		inst->smi_regs_ptr =
+		    ioremap(SMI_BASE_ADDRESS, sizeof(struct smi_register_map));
+		inst->cm_smi_regs_ptr =
+		    ioremap(CM_SMI_BASE_ADDRESS,
+			    sizeof(struct cm_smi_register_map));
+		inst->smi_regs_busaddr = SMI_BASE_ADDRESS + (0x7e000000 -
+			BCM2708_PERI_BASE);
+		inst->clock_source = 6;
+		inst->clock_divisor = 8;
+	}
+
+	/* Setup DMA channel if used */
+	if (USE_DMA) {
```

Comment:
It was helpful while writing the DMA stuff to be able to turn it on and off. Looks like it's all working now, so time for that to go.